File: Packages/com.github.homuler.mediapipe/Runtime/Scripts/External/Protobuf.cs
// Copyright (c) 2021 homuler
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

namespace Mediapipe
{
  public static class Protobuf
  {
    public delegate void LogHandler(int level, string filename, int line, string message);
    public static readonly LogHandler DefaultLogHandler = LogProtobufMessage;

    public static void SetLogHandler(LogHandler logHandler)
    {
      UnsafeNativeMethods.google_protobuf__SetLogHandler__PF(logHandler).Assert();
    }

    /// <summary>
    ///   Reset the <see cref="LogHandler" />.
    ///   If <see cref="SetLogHandler" /> is called, this method should be called before the program exits.
    /// </summary>
    public static void ResetLogHandler()
    {
      UnsafeNativeMethods.google_protobuf__ResetLogHandler().Assert();
    }

    [AOT.MonoPInvokeCallback(typeof(LogHandler))]
    private static void LogProtobufMessage(int level, string filename, int line, string message)
    {
      switch (level)
      {
        case 1:
          {
            UnityEngine.Debug.LogWarning($"[libprotobuf WARNING {filename}:{line}] {message}");
            return;
          }
        case 2:
          {
            UnityEngine.Debug.LogError($"[libprotobuf ERROR {filename}:{line}] {message}");
            return;
          }
        case 3:
          {
            UnityEngine.Debug.LogError($"[libprotobuf FATAL {filename}:{line}] {message}");
            return;
          }
        default:
          {
            UnityEngine.Debug.Log($"[libprotobuf INFO {filename}:{line}] {message}");
            return;
          }
      }
    }
  }
}

File: mediapipe_api/external/protobuf.cc
// Copyright (c) 2021 homuler
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

#include "mediapipe_api/external/protobuf.h"

#include <functional>
#include <string>

#include "google/protobuf/stubs/logging.h"

using google::protobuf::LogLevel;

namespace {
LogHandler* logHandler;
google::protobuf::LogHandler* defaultLogHandler;
}

void HandleProtobufLog(LogLevel level, const char* filename, int line, const std::string& message) { logHandler(level, filename, line, message.c_str()); }

MpReturnCode google_protobuf__SetLogHandler__PF(LogHandler* handler) {
  TRY
    logHandler = handler;
    auto prevLogHandler = google::protobuf::SetLogHandler(&HandleProtobufLog);

    if (defaultLogHandler == nullptr) {
      defaultLogHandler = prevLogHandler;
    }
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

MpReturnCode google_protobuf__ResetLogHandler() {
  TRY
    if (logHandler) {
      google::protobuf::SetLogHandler(defaultLogHandler);
    }
    logHandler = nullptr;
    RETURN_CODE(MpReturnCode::Success);
  CATCH_EXCEPTION
}

void mp_api_SerializedProtoArray__delete(mp_api::SerializedProto* serialized_proto_vector_data, int size) {
  auto serialized_proto = serialized_proto_vector_data;
  for (auto i = 0; i < size; ++i) {
    delete (serialized_proto++)->str;
  }
  delete[] serialized_proto_vector_data;
}
